rfcs/fast-n-furious.md

```diff
@@ -0,0 +1,354 @@
+
+- OXC parser, transformer & resolver were used (invoked from JS via N-API)
+- Some code transformations still occurred in the Node.js environment
+- Code was evaluated in the Node.js environment
+- `export *` statements were not supported
+
+The results in the real product were promising:
+
+- Current transformer (Linaria v3) - 37.9s
+- POC transformer - 4.6s (8x faster)
+
+The POC was not completed, but it demonstrated the potential to make the compiler faster.
+
+### Experiment B (made by [@anber](https://github.com/anber))
+
+> Ratio: 0% JS / 100% Rust
+
+Based on some of my findings, [@anber](https://github.com/anber) created another POC that passed a set of tests from `wyw-in-js` and also implemented injectable transformers.
+
+- OXC parser, transformer & resolver were used (invoked from Rust)
+- All code transformations occurred in the Rust environment
+- Code was evaluated in the Deno environment
+- No cache was implemented
+
+The results on his codebase were:  
+- Current CLI (`wyw-in-js`) - 1595ms
+- POC CLI - 224ms (7x faster)
+
+This POC also remained as a POC.
+
+## Fast & Furious 🏎️
+
+Our objective is to significantly enhance the compiler's speed, aiming for at least a 10x improvement.
+
+### Changes in tooling
+
+We will adopt Rust as the primary language for the next-gen compiler and use [`napi-rs`](https://github.com/napi-rs/napi-rs) to create Node.js bindings.
+
+#### Parser & transformer
+
+We will replace Babel with the OXC parser & transformer, which is the fastest available, being 3x faster than SWC and 20x-70x faster than Babel.
+
+- https://github.com/oxc-project/bench-javascript-parser-written-in-rust
+- https://github.com/oxc-project/bench-transformer
+
+#### Resolver
+
+A fast resolver is also necessary. While we will still offer the option to use a resolver from a bundler (e.g., Webpack, Vite, etc.), we will primarily use the OXC resolver, which is 28x faster than Webpack's resolver.
+
+- https://github.com/oxc-project/bench-resolver
+
+#### Evaluation
+
+We will experiment with using the Deno/Bun runtime for code execution, which should be faster than Node.js.
+
+> Note: Initial experiments indicated that Node.js was faster for running code than Deno.
+
+### Changes in approach
+
+We will also optimize the steps in our approach to further accelerate the compiler.
+
+```mermaid
+graph TD
+    A[prepare `input.ts`] --> B[collect imports, tree-shake, transform `input.ts`]
+    B --> D[resolve imports in `input.ts`]
+    D --> E[evaluate _bundle_]
+    D --> C2[collect imports, tree-shake, transform `theme.ts`]
+    C2 --> C3[resolve imports in `theme.ts`]
+    C3 --> E
+```
+
+#### Strict mode (performance mode)
+
+We will introduce a strict mode that disables certain features to enhance compiler speed. For instance, this mode will not support `export *` and CommonJS modules.
+
+**This mode will be implemented first.**
```

Comment:
Might want to make it clear that non-performant syntaxes like `export *` will be supported eventually?